cmd/occollector/app/builder/pipelines_builder.go

```diff
@@ -0,0 +1,164 @@
+)
+
+// builtProcessor is a processor that is built based on a config.
+// It can have a trace and/or a metrics consumer.
+type builtProcessor struct {
+	tc consumer.TraceConsumer
+	mc consumer.MetricsConsumer
+}
+
+// PipelineProcessors is a map of entry-point processors created from pipeline configs.
+// Each element of the map points to the first processor of the pipeline.
+type PipelineProcessors map[*configmodels.Pipeline]*builtProcessor
+
+// PipelinesBuilder builds pipelines from config.
+type PipelinesBuilder struct {
+	logger    *zap.Logger
+	config    *configmodels.ConfigV2
+	exporters Exporters
+}
+
+// NewPipelinesBuilder creates a new PipelinesBuilder. Requires exporters to be already
+// built via ExportersBuilder. Call Build() on the returned value.
+func NewPipelinesBuilder(
+	logger *zap.Logger,
+	config *configmodels.ConfigV2,
+	exporters Exporters,
+) *PipelinesBuilder {
+	return &PipelinesBuilder{logger, config, exporters}
+}
+
+// Build pipeline processors from config.
+func (eb *PipelinesBuilder) Build() (PipelineProcessors, error) {
+	pipelineProcessors := make(PipelineProcessors)
+
+	for _, pipeline := range eb.config.Pipelines {
+		firstProcessor, err := eb.buildPipeline(pipeline)
+		if err != nil {
+			return nil, err
+		}
+		pipelineProcessors[pipeline] = firstProcessor
+	}
+
+	return pipelineProcessors, nil
+}
+
+// Builds a pipeline of processors. Returns the first processor in the pipeline.
+// The last processor in the pipeline will be plugged to fan out the data into exporters
+// that are configured for this pipeline.
+func (eb *PipelinesBuilder) buildPipeline(
+	pipelineCfg *configmodels.Pipeline,
+) (*builtProcessor, error) {
+
+	// Build the pipeline backwards.
+
+	// First create a consumer junction point that fans out the data to all exporters.
+	var tc consumer.TraceConsumer
+	var mc consumer.MetricsConsumer
+
+	switch pipelineCfg.InputType {
+	case configmodels.TracesDataType:
+		tc = eb.buildFanoutExportersTraceConsumer(pipelineCfg.Exporters)
+	case configmodels.MetricsDataType:
+		mc = eb.buildFanoutExportersMetricsConsumer(pipelineCfg.Exporters)
+	}
+
+	// Now build the processors backwards, starting from the last one.
+	// The last processor points to consumer which fans out to exporters, then
+	// the processor itself becomes a consumer for the one that precedes it in
+	// in the pipeline and so on.
+	for i := len(pipelineCfg.Processors) - 1; i >= 0; i-- {
+		procName := pipelineCfg.Processors[i]
+		procCfg := eb.config.Processors[procName]
+
+		factory := factories.GetProcessorFactory(procCfg.Type())
+
+		// This processor must point to the next consumer and then
+		// it becomes the next for the previous one (previous in the pipeline,
+		// which we will build in the next loop iteration).
+		var err error
+		switch pipelineCfg.InputType {
+		case configmodels.TracesDataType:
+			tc, err = factory.CreateTraceProcessor(tc, procCfg)
```

Comment:
Just occurred to me that we may want to pass the logger to the factories so the instance have access to it. It is unfortunate that the logger is an instance, it will require us to pass it down some way. That said let's not bring it on this PR.

Reply:
Sure. We can add it later.